cmake/patch_llvm.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python3
+
+"""
+Script to apply a set of patches to llvm-project sources.
+"""
+
+import argparse
+import os
+import subprocess
+import sys
+
+
+def main():
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "patchdir",
+        help="Set of patches to apply. This should be a directory containing one or more ordered *.patch files.",
+    )
+    parser.add_argument(
+        "--llvm_dir",
+        help="Directory of the llvm-project git checkout, if not the current directory.",
+    )
+    parser.add_argument(
+        "--method",
+        choices=["am", "apply"],
+        default="apply",
+        help="Git command to use. git am will add each patch as a commit, whereas git apply will leave patched changes staged.",
+    )
+    parser.add_argument(
+        "--reset",
+        help="Clean and reset the repo to a specified commit before patching.",
+    )
+    parser.add_argument(
+        "--restore_on_fail",
+        action="store_true",
+        help="If a patch in a series cannot be applied, restore the original state instead of leaving patches missing. Return code will be 2 instead of 1.",
+    )
+    args = parser.parse_args()
+
+    if args.reset:
+        reset_args = ["git", "reset", "--quiet", "--hard"]
+        subprocess.check_output(reset_args, cwd=args.llvm_dir)
+        clean_args = ["git", "clean", "--quiet", "--force", "-dx"]
+        subprocess.check_output(clean_args, cwd=args.llvm_dir)
+
+    patch_names = []
+    for patch_name in os.listdir(args.patchdir):
+        if patch_name.endswith(".patch"):
+            patch_names.append(patch_name)
```

Comment:
Here, and onwards: 
The code currently extracts the file names only to later append them to the folder name and use that, multiple times. 
What we actually want is to work with file paths: 
Use glob to get all the relevant files, as we'd simply do in the shell : *.patch. 
Then we use file paths everywhere, and in the single place we just want the names, we simply do path.name().stem(). 

If I understand correctly, using pathlib 
pathlib.Path(args.patchdir).glob("*.patch")
would return an iterable with the relevant paths.

Reply:
Done, working with path objects does work out cleaner.